rules/copy_file_private.bzl

```diff
@@ -0,0 +1,121 @@
+# Copyright 2019 The Bazel Authors. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Implementation of copy_file and copy_xfile rules.
+
+These rules copy a file to another location using Bash (on Linux/macOS) or
+cmd.exe (on Windows). 'copy_xfile' marks the resulting file executable,
+'copy_file' does not.
+"""
+
+def _common_impl(ctx, is_executable):
+    if ctx.attr.is_windows:
+        # Most Windows binaries built with MSVC use a certain argument quoting
+        # scheme. Bazel uses that scheme too to quote arguments. However,
+        # cmd.exe uses different semantics, so Bazel's quoting is wrong here.
+        # To fix that we write the command to a .bat file so no command line
+        # quoting or escaping is required.
+        bat = ctx.actions.declare_file(ctx.label.name + "-cmd.bat")
+        ctx.actions.write(
+            output = bat,
+            content = "@copy /Y \"%s\" \"%s\" >NUL" % (
```

Comment:
lib/shell.bzl shell.escape on the args?

Reply:
I see no such function: https://github.com/bazelbuild/bazel-skylib/blob/master/lib/shell.bzl

Which one do you mean?

Reply:
Sorry, `shell.quote`

Reply:
Unfortunately that uses Bash quoting syntax, which doesn't apply for cmd.exe

Reply:
Perhaps a comment to that effect?

Reply:
Done.